mapillm/exact_matching.py

```diff
@@ -0,0 +1,131 @@
+        all_values = self._get_prop(all_values, property)
+        return sum(all_values) / len(all_values)
+   
+    def k_vals(self, input_param:dict, k:int):
+        num_values = 0
+        for key in input_param:
+            num_values += len(input_param[key])
+        return num_values >= k
+
+    def get_best_matches(self, formula, desired_prop, space_group_num=None, density=None, lattice=None, k=10):
+        """
+        Get the best matches following the scoring rubric.
+        """
+        N_params = {}
+
+        score_criteria = [
+            # Score 3 criteria
+            [
+                {"formula": formula, "lattice": lattice, "space_group_num": space_group_num, "desired_prop": desired_prop},
+                {"formula": formula, "lattice": lattice, "density": density, "desired_prop": desired_prop},
+                {"formula": formula, "space_group_num": space_group_num, "density": density, "desired_prop": desired_prop}
+            ],
+            # Score 2 criteria
+            [
+                {"formula": formula, "lattice": lattice, "desired_prop": desired_prop},
+                {"formula": formula, "space_group_num": space_group_num, "desired_prop": desired_prop},
+                {"formula": formula, "density": density, "desired_prop": desired_prop}
+            ],
+            # Score 1 criteria
+            [{"formula": formula, "desired_prop": desired_prop}]
+        ]
+        
+        for score, criteria in enumerate(score_criteria, start=1):
+            props = []
+            for criterion in criteria:
+                criterion["one_only"] = False  # Apply common property
+                docs = self.query_material_property(**criterion)
+                if docs:
+                    props += self._get_prop(docs, desired_prop)
+            N_params[score] = props
+            
+            if self.k_vals(N_params, k):
+                return N_params
+
+        return N_params
+
+    def get_weighted_average_of_neighbors(self, input_params:dict):
+        required_params = ["formula", "desired_prop"]
+        if not all([param in input_params for param in required_params]):
+            raise ValueError(f"Missing required parameters: {required_params}")
+        neighbors = self.get_best_matches(**input_params)
+        return self.average_neighbors(neighbors)
+    
+    def get_prop(self, formula, desired_prop, space_group_num=None, density=None, lattice=None, one_only=True):
+        docs = self.query_material_property(formula, desired_prop, space_group_num, density, lattice, one_only=False)
+        all_props = self._get_prop(docs, desired_prop)
+        if not docs:
+            print ("Warning: no matches found. Returning the average of closet matches.")
+            nearest_neighbor_avg = self.get_weighted_average_of_neighbors({"formula": formula, "desired_prop": desired_prop, "space_group_num": space_group_num, "density": density, "lattice": lattice})
```

Comment:
We need to compare this average pooling scheme with the ICL to decide which one to keep. But I think it's ok to keep that for now